projects/mci/test/test_mci_broadcast.c

```diff
@@ -0,0 +1,375 @@
+  LOG_DEBUG("TRANSMITTING IN LOOPBACK ...\n");
+  if (storage->rx_cb != NULL) {
+    LOG_DEBUG("CALLED_RECIEVER\n");
+    storage->rx_cb(id, extended, data, dlc, storage->context);
+  }
+  return STATUS_CODE_OK;
+}
+
+static void prv_rx_handler(uint32_t id, bool extended, uint64_t data, size_t dlc, void *context) {
+  GenericCanMcp2515 *gcmcp = context;
+  for (size_t i = 0; i < NUM_GENERIC_CAN_RX_HANDLERS; i++) {
+    if (gcmcp->base.rx_storage[i].rx_handler != NULL &&
+        (id & gcmcp->base.rx_storage[i].mask) == gcmcp->base.rx_storage[i].filter) {
+      const GenericCanMsg msg = {
+        .id = id,
+        .extended = extended,
+        .data = data,
+        .dlc = dlc,
+      };
+      gcmcp->base.rx_storage[i].rx_handler(&msg, gcmcp->base.rx_storage[i].context);
+      break;
+    }
+  }
+}
+
+static void prv_setup_motor_can(void) {
+  Mcp2515Settings mcp2515_settings = {
+    .spi_port = SPI_PORT_2,
+    .spi_baudrate = 6000000,
+    .mosi = { .port = GPIO_PORT_B, 15 },
+    .miso = { .port = GPIO_PORT_B, 14 },
+    .sclk = { .port = GPIO_PORT_B, 13 },
+    .cs = { .port = GPIO_PORT_B, 12 },
+    .int_pin = { .port = GPIO_PORT_A, 8 },
+
+    .can_bitrate = MCP2515_BITRATE_500KBPS,
+    .loopback = true,
+  };
+
+  // Will throw unimplemented errors on x86
+  generic_can_mcp2515_init(&s_motor_can, &mcp2515_settings);
+  // On x86 the rx handler doesn't get registered either :|
+  mcp2515_register_cbs(s_motor_can.mcp2515, prv_rx_handler, NULL, &s_motor_can);
+}
+
+static void prv_setup_system_can() {
+  CanSettings can_settings = {
+    .device_id = SYSTEM_CAN_DEVICE_MOTOR_CONTROLLER,
+    .bitrate = CAN_HW_BITRATE_500KBPS,
+    .rx_event = MCI_CAN_EVENT_RX,
+    .tx_event = MCI_CAN_EVENT_TX,
+    .fault_event = MCI_CAN_EVENT_FAULT,
+    .tx = { GPIO_PORT_A, 12 },
+    .rx = { GPIO_PORT_A, 11 },
+    .loopback = true,
+  };
+
+  can_init(&s_can_storage, &can_settings);
+}
+
+void setup_test(void) {
+  event_queue_init();
+  interrupt_init();
+  soft_timer_init();
+
+  prv_setup_system_can();
+  prv_setup_motor_can();
+
+  TEST_ASSERT_OK(can_register_rx_handler(SYSTEM_CAN_MESSAGE_MOTOR_CONTROLLER_VC,
+                                         prv_handle_bus_measurement, NULL));
+  TEST_ASSERT_OK(
+      can_register_rx_handler(SYSTEM_CAN_MESSAGE_MOTOR_VELOCITY, prv_handle_velocity, NULL));
+}
+
+void teardown_test(void) {
+  s_expect_velocity = false;
+  s_expect_bus_measurement = false;
+}
+
+// lv - left velocity
+// rv - right velocity
+// lb - left bus measurement
+// rb - right bus measurement
+
+// Test 1: lb rv lv rb (check 2 output)
+void test_all_measurements_lb_rv_lv_rb() {
+  // motor_can tx and rx should happen immedietly
+  MotorControllerBroadcastStorage broadcast_storage = { 0 };
+  mci_broadcast_init(&broadcast_storage, &s_broadcast_settings);
+
+  s_expect_velocity = true;
+  s_expect_bus_measurement = true;
+
+  prv_send_measurements(LEFT_MOTOR_CONTROLLER, TEST_MCI_BUS_MEASUREMENT_MESSAGE, false);
```

Comment:
I like your tests, they're rigorous, and granular. I usually prefer putting the inputs and the outputs in the test itself. I think it'd be neat if you could have your prv_send_measurements function to take the actual measurement value as an input, then in your callbacks you could set a global variable to capture the sent value, then make the assertions at the end of your test instead of the callbacks. I think that would make your tests easier to understand. Cuz when I'm looking at your tests now I'm like, oh two messages got TX and RX'd, now what? then I had to take a look at your callback to see if the assertions were being made. Putting your final assertions in your test makes them flow better :)